### PR TITLE
Versioning policy.

### PR DESCRIPTION
**Pull request by monopole**
for #746

### CODE REVIEW

**Comment by Liujingfang1** [comment]
I also write a doc about kustomize 2.0.0 https://github.com/kubernetes-sigs/kustomize/pull/752. There are some overlaps with this one.

**Comment by Liujingfang1** [comment]
Got it.

**Comment by monopole** [comment]
Would love to know if there's some apimachinery magic that would make this significantly better for the user.

File: docs/versionPolicy.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
+# Versioning
+
+Running `kustomize` means one is running a
```
Comment (Liujingfang1) [inline docs/versionPolicy.md]: The program version and kustomization file version is related. The descriptions here and structure below make me feel that they are separate. 

Move some later session like 
```
 Kustomization Field Change Policy
```
outside of `Kustomization type versioning` may help better understanding.

File: docs/versionPolicy.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+version tag that aspires to [semantic versioning].
+
+When enough changes have accumulated to
+(subjectively) warrant a new release, a [release process]
+is followed, and the fields in the version number
+are bumped per semver.
+
+## Kustomization type versioning
+
+At the time of writing, a [kustomization] file is just
```
Comment (Liujingfang1) [inline docs/versionPolicy.md]: How about changing `At the time of writing` to `Until 2.0.0`.

File: docs/versionPolicy.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+non-Go struct (i.e. CRD) representation of
+the kustomization type, a code
+generation stage in building (no more `go get`),
+and Go code associating field deprecations
+with versions - perhaps a hardcoded string map.
+Just to change a command's error message.
+
+The latter seems enough for now, and is already policy.
+
+When/if the K struct graduates to some kind of API status,
```
Comment (Liujingfang1) [inline docs/versionPolicy.md]: Change to `K` to `Kustomization`

File: docs/versionPolicy.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+[beta-level rules]: https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#alpha-beta-and-stable-versions
+[api versioning]: https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#alpha-beta-and-stable-versions
+[adapt]: https://github.com/kubernetes-sigs/kustomize/blob/master/pkg/types/kustomization.go#L166
+[special]: https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#resources
+[k8s API]: https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#resources
+[release process]: ../build/README.md
+[kustomization]: glossary.md#kustomization
+[`kind`]: https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#types-kinds
+[`apiVersion`]: https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning
+[Semantic Versioning]: https://semver.org
```
Comment (Liujingfang1) [inline docs/versionPolicy.md]: excellent links!

File: docs/versionPolicy.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+generation stage in building (no more `go get`),
+and Go code associating field deprecations
+with versions - perhaps a hardcoded string map.
+Just to change a command's error message.
+
+The latter seems enough for now, and is already policy.
+
+When/if the K struct graduates to some kind of API status,
+whatever it looks like at that moment can be locked
+into `/v1beta1`.
```
Comment (Liujingfang1) [inline docs/versionPolicy.md]: Any policy of moving it to `v1`?

File: docs/versionPolicy.md (hunk 1 above)
Comment (monopole) [reply docs/versionPolicy.md]: you're right!  tried to fix.

File: docs/versionPolicy.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+a YAML file that can be successfully parsed into a
+particular Go struct defined in the `kustomize`
+binary.
+
+This struct does not have a version number, which is the
+same as saying that its version number matches
+the program's version number, since it's compiled in.
+
+#### Kustomization Field Change Policy
+- A field's meaning cannot be changed.
```
[the remark this answers is not in the document]
Comment (monopole) [reply docs/versionPolicy.md]: yes, adjusted.

File: docs/versionPolicy.md [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+### The `edit fix` command
+
+This `kustomize` command reads a Kustomization
+file, translates _deprecated_ fields to new
+fields, and writes it out again in the latest
+format.
+
+This is a type version upgrade mechanism.  There
+is no downgrade capability, as there's no use case
```
Comment (mengqiy) [inline docs/versionPolicy.md]: `edit fix` command doesn't provide ability to go back.
But if the users use GitOps in their workflow, they can always go back.
We might write it down if you think it's useful.

File: docs/versionPolicy.md [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+field version tag (e.g. `1.0.11`) that aspires to
+[semantic versioning].
+
+When enough changes have accumulated to
+(subjectively) warrant a new release,
+a [release process] is followed, and the
+fields in the version number are bumped
+per semver.
+
+## Kustomization file versioning
```
Comment (mengqiy) [inline docs/versionPolicy.md]: This section is very useful explaining why doing versioning in `kustomization.yaml` may not be useful at the time being :)

File: docs/versionPolicy.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
+# Versioning
+
+Running `kustomize` means one is running a
+particular version of a program, reading a
+particular version of a [kustomization] file.
+
+## Program versioning
+
+The command `kustomize version` prints a three field
```
[the remark this answers is not in the document]
Comment (monopole) [reply docs/versionPolicy.md]: done

File: docs/versionPolicy.md (hunk 3 above)
Comment (monopole) [reply docs/versionPolicy.md]: done

File: docs/versionPolicy.md (hunk 7 above)
Comment (monopole) [reply docs/versionPolicy.md]: updated, thanks!

File: docs/versionPolicy.md (hunk 8 above)
Comment (monopole) [reply docs/versionPolicy.md]: thanks

**Comment by monopole** [comment]
OK, ready for another pass if anyone interested.

File: docs/versionPolicy.md (hunk 2 above)
Comment (monopole) [reply docs/versionPolicy.md]: done!

**Pull request merged**